### PR TITLE
Cosmos DB compatible search

### PR DESCRIPTION
**Pull request by algattik**
Escape special characters one by one when searching Mongo DB, rather than using Pattern.quote(), for Cosmos DB compatibility. (Azure Cosmos DB can be configured with a Mongo DB compatible API, but it supports ECMAScript regexes instead of PCRE).

Fixes AbsaOSS/spline#166.

### CODE REVIEW

**Comment by wajda** [comment]
Thank you @algattik, we really appreciate your contribution!
Could you please add a unit test that tests your function in different boundary conditions?

Also, there are a few rather minor things that I'd like to outline in the comments below. It would be cool if you fix them as well while you are on it.

Just for information, we are now working on the Contribution guide applied not only for Spline but for all ABSA OSS projects, that aims to unify coding standards and make the contribution process as smooth as possible for both externals as well as our internal contributors. It will be published shortly.

File: persistence/mongo/src/main/scala/za/co/absa/spline/persistence/mongo/dao/BaselineLineageDAO.scala [hunk 1]
```diff
@@ -48,6 +47,8 @@ abstract class BaselineLineageDAO extends VersionedLineageDAO with Logging {
   protected lazy val dataLineageCollection: DBCollection = getMongoCollectionForComponent(Component.Root)
   protected lazy val operationCollection: DBCollection = getMongoCollectionForComponent(Component.Operation)
 
+  private final val NON_WORD_CHAR = "\\W".r
```
Comment (wajda) [inline persistence/mongo/src/main/scala/za/co/absa/spline/persistence/mongo/dao/BaselineLineageDAO.scala]: Consider placing constants into a companion object.

File: persistence/mongo/src/main/scala/za/co/absa/spline/persistence/mongo/dao/BaselineLineageDAO.scala [hunk 2]
```diff
@@ -235,13 +236,25 @@ abstract class BaselineLineageDAO extends VersionedLineageDAO with Logging {
...
 
+  /**
+    * Escape special characters in a regular expression.
+    * Does not use Pattern.quote() for Cosmos DB compatibility.
+    * @see <a href="https://github.com/AbsaOSS/spline/issues/166">issue #166</a>
+    *
+    * @param text A string
+    * @return A regular expression searching for the exact literal text string
+    */
+  protected def quoteSafely(text: String): String = {
```
Comment (wajda) [inline persistence/mongo/src/main/scala/za/co/absa/spline/persistence/mongo/dao/BaselineLineageDAO.scala]: Is there any reason for the method being `protected` instead of `private` ?

File: persistence/mongo/src/main/scala/za/co/absa/spline/persistence/mongo/dao/BaselineLineageDAO.scala [hunk 3]
```diff
@@ -235,13 +236,25 @@ abstract class BaselineLineageDAO extends VersionedLineageDAO with Logging {
...
+    return NON_WORD_CHAR.replaceAllIn(text, "\\\\" + _.matched)
```
Comment (wajda) [inline persistence/mongo/src/main/scala/za/co/absa/spline/persistence/mongo/dao/BaselineLineageDAO.scala]: Minor: generally we prefer string interpolation over string concatenation. Also consider using triple-quotes to avoid escaping back-slashes.

File: persistence/mongo/src/main/scala/za/co/absa/spline/persistence/mongo/dao/BaselineLineageDAO.scala (hunk 2 above)
Comment (algattik) [reply persistence/mongo/src/main/scala/za/co/absa/spline/persistence/mongo/dao/BaselineLineageDAO.scala]: overengineering probably :) thought it could be an extension point for other storage implementations

File: persistence/mongo/src/main/scala/za/co/absa/spline/persistence/mongo/dao/BaselineLineageDAO.scala (hunk 3 above)
Comment (algattik) [reply persistence/mongo/src/main/scala/za/co/absa/spline/persistence/mongo/dao/BaselineLineageDAO.scala]: Unfortunately triple-quotes don't reduce the number of required backslashes.

**Comment by algattik** [comment]
Thanks a lot @wajda, great feedback! I've improved the code added a unit test.

File: persistence/mongo/src/main/scala/za/co/absa/spline/persistence/mongo/dao/BaselineLineageDAO.scala [hunk 4]
```diff
@@ -47,7 +47,9 @@ abstract class BaselineLineageDAO extends VersionedLineageDAO with Logging {
   protected lazy val dataLineageCollection: DBCollection = getMongoCollectionForComponent(Component.Root)
   protected lazy val operationCollection: DBCollection = getMongoCollectionForComponent(Component.Operation)
 
-  private final val NON_WORD_CHAR = "\\W".r
+  object BaselineLineageDao {
```
Comment (wajda) [inline persistence/mongo/src/main/scala/za/co/absa/spline/persistence/mongo/dao/BaselineLineageDAO.scala]: No, I meant place it into the companion object of the `BaselineLineageDAO` class. It already exists at the bottom of the file.

File: persistence/mongo/src/main/scala/za/co/absa/spline/persistence/mongo/dao/BaselineLineageDAO.scala (hunk 3 above)
Comment (wajda) [reply persistence/mongo/src/main/scala/za/co/absa/spline/persistence/mongo/dao/BaselineLineageDAO.scala]: Why? There is no escape characters in Scala's multi-line string literal.
```scala
scala> "\\\\"
res0: String = \\

scala> """\\\\"""
res1: String = \\\\

scala> "\\\\" == """\\"""
res2: Boolean = true
```

File: persistence/mongo/src/main/scala/za/co/absa/spline/persistence/mongo/dao/BaselineLineageDAO.scala (hunk 4 above)
Comment (algattik) [reply persistence/mongo/src/main/scala/za/co/absa/spline/persistence/mongo/dao/BaselineLineageDAO.scala]: Got it. Fixed.

File: persistence/mongo/src/main/scala/za/co/absa/spline/persistence/mongo/dao/BaselineLineageDAO.scala (hunk 3 above)
Comment (wajda) [reply persistence/mongo/src/main/scala/za/co/absa/spline/persistence/mongo/dao/BaselineLineageDAO.scala]: Interesting :) Thanks.

**Comment by wajda** [comment]
The method `quoteSafely` seems to fail on the input with backslashes. It needs more tests.
```scala
scala> NON_WORD_CHAR.replaceAllIn("this is a text with backslash \\", m => s"\\\\${m.matched}")
java.lang.IllegalArgumentException: character to be escaped is missing
  at java.util.regex.Matcher.appendReplacement(Matcher.java:809)
  at scala.util.matching.Regex$Replacement.replace(Regex.scala:900)
  at scala.util.matching.Regex$Replacement.replace$(Regex.scala:900)
  at scala.util.matching.Regex$MatchIterator$$anon$4.replace(Regex.scala:878)
  at scala.util.matching.Regex.$anonfun$replaceAllIn$1(Regex.scala:508)
  at scala.collection.Iterator.foreach(Iterator.scala:941)
  at scala.collection.Iterator.foreach$(Iterator.scala:941)
  at scala.collection.AbstractIterator.foreach(Iterator.scala:1429)
  at scala.util.matching.Regex.replaceAllIn(Regex.scala:508)
  ... 28 elided
```

**Comment by algattik** [comment]
Fixed replacement for special chars, thanks
